app/models/service.rb

```diff
@@ -69,55 +70,28 @@ def request_type
     'service_reconfigure'
   end
 
-  def root_service
-    result = self
-    until result.parent_service.nil?
-      result = result.parent_service
-    end
-    result
-  end
-
-  def ancestors
-    result = []
-    node = self
-
-    while (node = node.parent_service)
-      result << node
-    end
-
-    result
-  end
-
-  def direct_service_children
-    services
-  end
+  alias root_service root
+  alias services children
+  alias direct_service_children children
 
   def indirect_service_children
-    direct_service_children.collect { |s| s.direct_service_children + s.indirect_service_children }.flatten.compact
-  end
-
-  def descendants
-    all_service_children
-  end
-
-  def subtree
-    all_service_children + [self]
+    descendants(:from_depth => 2)
   end
+  Vmdb::Deprecation.deprecate_methods(self, :indirect_service_children)
 
-  def all_service_children
-    direct_service_children + indirect_service_children
-  end
+  alias all_service_children descendants
 
   def indirect_vms
-    all_service_children.collect(&:direct_vms).flatten.compact
+    MiqPreloader.preload_and_map(indirect_service_children, :direct_vms)
   end
+  Vmdb::Deprecation.deprecate_methods(self, :indirect_vms)
 
   def direct_vms
-    service_resources.collect { |sr| sr.resource.kind_of?(VmOrTemplate) ? sr.resource : nil }.flatten.compact
+    service_resources.where(:resource_type => 'VmOrTemplate').collect(&:resource)
```

Comment:
this is still an n+1, but it only does it for the vms (vs before it did it for every resource that existed)